[JSC] Use cloning for Object.assign with empty object

Constellation opened the pull request:
#### 11aee4671a9e0291cec87cf0ebeb615d3f6ac51c
<pre>
[JSC] Use cloning for Object.assign with empty object
<a href="https://bugs.webkit.org/show_bug.cgi?id=267239">https://bugs.webkit.org/show_bug.cgi?id=267239</a>
<a href="https://rdar.apple.com/120667395">rdar://120667395</a>

Reviewed by Justin Michaud.

This patch implements fast cloning path for `Object.assign({}, source)` case.
When conditions are met, we clone Butterfly in source, set it to the empty target object,
and use the source&apos;s Structure for target. This completely avoids iterating all object properties.

* JSTests/stress/object-assign-clone.js: Added.
(shouldBe):
(shouldBeArray):
(shouldBe.Reflect.getOwnPropertyDescriptor):
(shouldBe.set get let):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::fastForEachPropertyWithSideEffectFreeFunctor): Deleted.
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::toPropertyDescriptor):
(JSC::defineProperties):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectCloneFast):
(JSC::objectAssignFast):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::Structure):
(JSC::Structure::nonPropertyTransitionSlow):
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::seenProperties const):
(JSC::Structure::bitFieldFlagsCantBeChangedWithoutTransition):
(JSC::Structure::transitionPropertyAttributes const):
(JSC::Structure::setTransitionPropertyAttributes):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::add):
(JSC::Structure::attributeChange):
(JSC::Structure::addOrReplacePropertyWithoutTransition):

Canonical link: <a href="https://commits.webkit.org/272794@main">https://commits.webkit.org/272794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36f8e92d286741ec4af2a04827f9cfe6c14fe191

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35011 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35742 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/29894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9045 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33576 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/10005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/29565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/8725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/29537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37073 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/28326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/30073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/29929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/34996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/33115 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/9015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/6952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/32850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/10716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/39573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/8317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4256 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/9659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -311,7 +311,7 @@ JSC_DEFINE_JIT_OPERATION(operationObjectAssignObject, void, (JSGlobalObject* glo
 
         // Do not clear since Vector::clear shrinks the backing store.
-        bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+        bool objectAssignFastSucceeded = objectAssignFast(globalObject, jsCast<JSFinalObject*>(target), source, properties, values);
```
Maybe, this is better?
```
    JSFinalObject* targetFinalObject = jsDynamicCast<JSFinalObject*>(target);
    bool targetCanPerformFastPut = targetFinalObject && target->canPerformFastPutInlineExcludingProto() && target->isStructureExtensible();
    ...
    bool objectAssignFastSucceeded = objectAssignFast(globalObject, targetFinalObject, source, properties, values);
```

Constellation replied:
A bit changed.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -344,7 +344,7 @@ JSC_DEFINE_JIT_OPERATION(operationObjectAssignUntyped, void, (JSGlobalObject* gl
         Vector<RefPtr<UniquedStringImpl>, 8> properties;
         MarkedArgumentBuffer values;
-        bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+        bool objectAssignFastSucceeded = objectAssignFast(globalObject, jsCast<JSFinalObject*>(target), source, properties, values);
```
Ditto

Constellation replied:
A bit changed.

hyjorc1 commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -387,7 +387,7 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
             }
 
-            bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+            bool objectAssignFastSucceeded = objectAssignFast(globalObject, jsCast<JSFinalObject*>(target), source, properties, values);
```
Ditto

Constellation replied:
Done.

justinmichaud commented on Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
```diff
@@ -63,7 +63,145 @@ ALWAYS_INLINE void objectAssignIndexedPropertiesFast(JSGlobalObject* globalObjec
+    if (!checkStrucure(sourceStructure))
+        return false;
+
+    if (targetStructure->inlineCapacity() != sourceStructure->inlineCapacity()) {
```
This is fine as long as targetStructure->inlineCapacity is greater than source, right?

Constellation replied:
No since we are using `sourceStructure` in the target. So we need to ensure that target object has the same inlineCapacity.

justinmichaud reviewed (approved):
r=me

Constellation committed:
[JSC] Use cloning for Object.assign with empty object
https://bugs.webkit.org/show_bug.cgi?id=267239
rdar://120667395

Reviewed by Justin Michaud.

This patch implements fast cloning path for `Object.assign({}, source)` case.
When conditions are met, we clone Butterfly in source, set it to the empty target object,
and use the source's Structure for target. This completely avoids iterating all object properties.

* JSTests/stress/object-assign-clone.js: Added.
(shouldBe):
(shouldBeArray):
(shouldBe.Reflect.getOwnPropertyDescriptor):
(shouldBe.set get let):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::fastForEachPropertyWithSideEffectFreeFunctor): Deleted.
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::toPropertyDescriptor):
(JSC::defineProperties):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectCloneFast):
(JSC::objectAssignFast):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::Structure):
(JSC::Structure::nonPropertyTransitionSlow):
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::seenProperties const):
(JSC::Structure::bitFieldFlagsCantBeChangedWithoutTransition):
(JSC::Structure::transitionPropertyAttributes const):
(JSC::Structure::setTransitionPropertyAttributes):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::add):
(JSC::Structure::attributeChange):
(JSC::Structure::addOrReplacePropertyWithoutTransition):

Canonical link: https://commits.webkit.org/272794@main
